src/coreclr/gc/unix/gcenv.unix.cpp

```diff
@@ -686,29 +702,31 @@ bool GCToOSInterface::VirtualDecommit(void* address, size_t size)
 //  true if it has succeeded, false if it has failed
 bool GCToOSInterface::VirtualReset(void * address, size_t size, bool unlock)
 {
-    int st;
-#if HAVE_MADV_FREE
+    int st = EINVAL;
+    int madviseFlags = 0;
+
+#ifdef MADV_DONTDUMP
+    // Do not include reset memory in coredump.
+    madviseFlags |= MADV_DONTDUMP;
+#endif
+
+#ifdef HAVE_MADV_FREE
     // Try to use MADV_FREE if supported. It tells the kernel that the application doesn't
     // need the pages in the range. Freeing the pages can be delayed until a memory pressure
     // occurs.
-    st = madvise(address, size, MADV_FREE);
-    if (st != 0)
+    madviseFlags |= MADV_FREE;
 #endif
+
+    if (madviseFlags != 0)
     {
-#if HAVE_POSIX_MADVISE
-        // In case the MADV_FREE is not supported, use MADV_DONTNEED
-        st = posix_madvise(address, size, MADV_DONTNEED);
-#else
-        // If we don't have posix_madvise, report failure
-        st = EINVAL;
-#endif
+        st = madvise(address, size, madviseFlags);
     }
 
-#ifdef MADV_DONTDUMP
-    if (st == 0)
+#if HAVE_POSIX_MADVISE
+    if (st != 0)
     {
-        // Do not include reset memory in coredump.
-        madvise(address, size, MADV_DONTDUMP);
+        // In case the MADV_FREE is not supported, use MADV_DONTNEED
+        st = posix_madvise(address, size, MADV_DONTNEED);
```

Comment:
We are missing the `MADV_DONTDUMP` for this case.

Reply:
Annoyingly `MADV_DONTDUMP` isn't supported for `posix_madvise()`, so we need to call `madvise()` again.

Reply:
i've just found an interesting comment in the posix_madvise doc:
> In glibc, this function is implemented using [madvise(2)](https://man7.org/linux/man-pages/man2/madvise.2.html).
       However, since glibc 2.6, POSIX_MADV_DONTNEED is treated as a no-
       op, because the corresponding [madvise(2)](https://man7.org/linux/man-pages/man2/madvise.2.html) value, MADV_DONTNEED,
       has destructive semantics.

Interestingly, we are passing `MADV_DONTNEED` instead of `POSIX_MADV_DONTNEED` to the `posix_madvise`. I don't know what is the actual result of calling `posix_madvise` in such case.

Reply:
> Interestingly, we are passing `MADV_DONTNEED` instead of `POSIX_MADV_DONTNEED` to the `posix_madvise`. I don't know what is the actual result of calling `posix_madvise` in such case.

These are both 4. I've updated to `POSIX_MADV_DONTNEED` to make it explicit.

> i've just found an interesting comment in the posix_madvise doc:
> 
> > In glibc, this function is implemented using [madvise(2)](https://man7.org/linux/man-pages/man2/madvise.2.html).
> > However, since glibc 2.6, POSIX_MADV_DONTNEED is treated as a no-
> > op, because the corresponding [madvise(2)](https://man7.org/linux/man-pages/man2/madvise.2.html) value, MADV_DONTNEED,
> > has destructive semantics.
>

Glibc 2.6 was introduced in 2007. Ubuntu 22.04 has Glibc 2.35.

Given that and all the ifdef protections, it feels very unlikely that calling `posix_madvise()` _after_ a failing `madvise()` would ever do anything.

I've updated so that `posix_madvise()` is only called if `MADV_DONTDUMP` does not exist.